### PR TITLE
[ENH] Add Information Gain Temporal Segmentation Estimator

### PR DESCRIPTION
**Pull request by lmmentel**
#### What does this implement/fix? Explain your changes.

Information Gain-based Temporal Segmentation.

Information Gain Temporal Segmentation (IGTS) is a method for segmenting
multivariate time series based off reducing the entropy in each segment [1]_.

The amount of entropy lost by the segmentations made is called the Information
Gain (IG). The aim is to find the segmentations that have the maximum information
gain for any number of segmentations.

References
----------
.. [1] Sadri, Amin, Yongli Ren, and Flora D. Salim.
    "Information gain-based metric for recognizing transitions in human activities.",
    Pervasive and Mobile Computing, 38, 92-109, (2017).
    https://www.sciencedirect.com/science/article/abs/pii/S1574119217300081

#### Does your contribution introduce a new dependency? If yes, which one?
NO.

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [x] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [x] I've added unit tests and made sure they pass locally.
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.

##### For new estimators
- [x] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

### CODE REVIEW

**Comment by lmmentel** [comment]
FYI @miraep8, @KatieBuc here is my take on implementing IGTS. Let me know if you have any thoughts or suggestions.

**Comment by lmmentel** [comment]
@fkiraly some test automation breaks with a logging statement that has an f-string (I think). Would be worth for you to ave a look at https://github.com/alan-turing-institute/sktime/actions/runs/3074385309/jobs/4967169980

**Review by fkiraly — CHANGES_REQUESTED** [review]
Could you kindly explain why the base tests break?
Would be great if you could look into it, what are the error messages coming from the test suite?

**Comment by lmmentel** [comment]
All the test are passing, I would be grateful if you could review the PR if you can so that we can move forward with it. 

I have since removed the offending logging statement. I have not investigated why it was failing tests but the error was coming from an `eval` statement.

**Comment by fkiraly** [comment]
> All the test are passing, I would be grateful if you could review the PR if you can so that we can move forward with it.

@lmmentel, I don't think that's a true statement.

You've added the following skips/excepts for tests in the test config file. Either these tests are failing, or the skips are unnecessary.

I an kindly asking you to explain these - these are tests of sklearn and test framework compatibility that every sktime estimator should pass.

```python
    "InformationGainSegmentation": [
        "test_constructor",
        "test_inheritance",
        "test_create_test_instance",
    ],
```

**Comment by lmmentel** [comment]
These test are a bit too constraining given that they test implementation rather than functionality. This estimator is compatible with sklearn interface.

**Comment by fkiraly** [comment]
> These test are a bit too constraining given that they test implementation rather than functionality. This estimator is compatible with sklearn interface.

Apologies, that's not a very clear statement to me, can you please explain?

The tests above test specific interface assumptions required for `sklearn` compatibility, failing them will typically make your estimators not `sklearn` compatible.

We need to be `sklearn` compatible even if we otherwise go with a custom interface.

**Comment by fkiraly** [comment]
I've added some explanations to the respective tests here:
https://github.com/sktime/sktime/pull/3555

to explain why these tests are important and how to comply with them.
Hope that is useful?

**Comment by lmmentel** [comment]
Not sure what you're asking for here. Are you saying that those tests have to included and pass or you want some more explanations from me?

**Comment by lmmentel** [comment]
FYI: similar tests are skipped for GGS https://github.com/sktime/sktime/blob/13239f38ec9ac8157f70aa2c70487c7978e38ac9/sktime/tests/_config.py#L110

**Comment by fkiraly** [comment]
> Not sure what you're asking for here. Are you saying that those tests have to included and pass or you want some more explanations from me?

Yes, at least one of the two, please.
GGS passes `test_constructor` which is required for much of `sklearn` compatibility, so it's less "bad" than this.

Preferably, all tests pass, although I'm happy with skipping `test_inheritance` until the base class question is solved.

File: sktime/annotation/igts.py [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+Gain (IG). The aim is to find the segmentations that have the maximum information
+gain for any number of segmentations.
+
+References
+----------
+.. [1] Sadri, Amin, Yongli Ren, and Flora D. Salim.
+    "Information gain-based metric for recognizing transitions in human activities.",
+    Pervasive and Mobile Computing, 38, 92-109, (2017).
+    https://www.sciencedirect.com/science/article/abs/pii/S1574119217300081
+
```
Comment (aiwalter) [inline sktime/annotation/igts.py:18]: this should go into class docstring if not there already

File: sktime/annotation/igts.py [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+"""
+
+# import logging
```
Comment (aiwalter) [inline sktime/annotation/igts.py:21]:
```suggestion
```

File: sktime/annotation/igts.py [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+import pandas as pd
+from attrs import asdict, define, field
+
+from sktime.base import BaseEstimator
+
+__all__ = ["InformationGainSegmentation"]
+__author__ = ["lmmentel"]
+
+
+# logger = logging.getLogger(__name__)
```
Comment (aiwalter) [inline sktime/annotation/igts.py:36]:
```suggestion
```

File: sktime/annotation/igts.py [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+            )
+
+            # logger.info(
+            #     f"BEST {ig_max=:.5f} for {best_candidate=} "
+            #     + f"current {current_change_points=}"
+            # )
+        return current_change_points
+
+
+class SegmentationMixin:
```
Comment (aiwalter) [inline sktime/annotation/igts.py:261]: should this not go into another file? Do you think its useful to be used by other algos?

File: sktime/annotation/igts.py [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+       IGTS does not work very well for univariate series but it can still be
+       used if the original univariate series are augmented by an extra feature
+       dimensions. A technique proposed in the paper [1]_ us to subtract the
+       series from it's largest element and append to the series.
+
+    Parameters
+    ----------
+    k_max: int, default=10
+        Maximum number of change points to find. The number of segments is thus k+1.
+
```
Comment (aiwalter) [inline sktime/annotation/igts.py:319]:
```suggestion
```

File: sktime/annotation/igts.py [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+    References
+    ----------
+    .. [1] Sadri, Amin, Yongli Ren, and Flora D. Salim.
+       "Information gain-based metric for recognizing transitions in human activities.",
+       Pervasive and Mobile Computing, 38, 92-109, (2017).
+       https://www.sciencedirect.com/science/article/abs/pii/S1574119217300081
+
+    Example
+    -------
```
Comment (aiwalter) [inline sktime/annotation/igts.py:349]:
```suggestion
    Examples
    --------
```

File: sktime/annotation/igts.py [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+    >>> from sktime.annotation.datagen import piecewise_normal_multivariate
+    >>> from sklearn.preprocessing import MinMaxScaler
+    >>> X = piecewise_normal_multivariate(lengths=[10, 10, 10, 10],
+    ... means=[[0.0, 1.0], [11.0, 10.0], [5.0, 3.0], [2.0, 2.0]],
+    ... variances=0.5)
```
Comment (aiwalter) [inline sktime/annotation/igts.py:354]:
```suggestion
    >>> X = piecewise_normal_multivariate(
    ... lengths=[10, 10, 10, 10],
    ... means=[[0.0, 1.0], [11.0, 10.0], [5.0, 3.0], [2.0, 2.0]],
    ... variances=0.5,
    ... )
```

File: sktime/annotation/igts.py [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+    >>> X_scaled = MinMaxScaler(feature_range=(0, 1)).fit_transform(X)
+    >>> from sktime.annotation.igts import InformationGainSegmentation
```
Comment (aiwalter) [inline sktime/annotation/igts.py:356]: would put all imports at top

File: sktime/annotation/igts.py [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+            Placeholder for compatibility with sklearn-api, not used, default=None.
+
+        Returns
+        -------
+        y_pred : array_like
+            1D array with predicted segmentation of the same size as the first
+            dimension of X. The numerical values represent distinct segments
+            labels for each of the data points.
+        """
+        return self.fit(X=X, y=y).predict(X=X, y=y)
```
Comment (aiwalter) [inline sktime/annotation/igts.py:434]: this should go and we can inherit from `from sklearn.base import BaseEstimator` which is implementing that?

File: sktime/annotation/tests/test_igts.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+"""Tests for igts module."""
+
+import numpy as np
+import pytest
+
+from sktime.annotation.igts import IGTS, InformationGainSegmentation, entropy
```
Comment (aiwalter) [inline sktime/annotation/tests/test_igts.py:7]: I am wondering why you can import `IGTS` and `entropy` here bcz you have in that file `__all__ = ["InformationGainSegmentation"]
`?

File: sktime/annotation/igts.py [hunk 11]
```diff
@@ -0,0 +1,486 @@
...
+            self.intermediate_results_.append(
+                ChangePointResult(
+                    k=k, score=ig_max, change_points=current_change_points.copy()
+                )
+            )
+
+            # logger.info(
+            #     f"BEST {ig_max=:.5f} for {best_candidate=} "
+            #     + f"current {current_change_points=}"
+            # )
```
Comment (aiwalter) [inline sktime/annotation/igts.py:257]:
```suggestion
```

File: sktime/annotation/igts.py [hunk 12]
```diff
@@ -0,0 +1,486 @@
...
+
+        for k in range(self.k_max):
+            ig_max = 0
+            # find a point which maximizes score
+            for candidate in self.get_candidates(n_samples, current_change_points):
+                try_change_points = {candidate}
+                try_change_points.update(current_change_points)
+                try_change_points = sorted(try_change_points)
+                ig = self.information_gain_score(X, try_change_points)
+                # logger.info(f"{ig=:.5f} for {candidate=} with {try_change_points=}")
```
Comment (aiwalter) [inline sktime/annotation/igts.py:241]:
```suggestion
```

**Review by aiwalter — COMMENTED** [review]
left some comments.

File: sktime/annotation/igts.py (hunk 4 above)
Comment (lmmentel) [reply sktime/annotation/igts.py:261]: Yes, that would be ultimate goal, however I would like to keep this PR small, and work on the changes in following PRs.

File: sktime/annotation/igts.py (hunk 9 above)
Comment (lmmentel) [reply sktime/annotation/igts.py:434]: Right now `BaseEstimator` does not have this method https://github.com/sktime/sktime/blob/main/sktime/base/_base.py and it's required to pass the tests.

File: sktime/annotation/tests/test_igts.py (hunk 10 above)
Comment (lmmentel) [reply sktime/annotation/tests/test_igts.py:7]: `__all__` is only relevant when you do `from foo import *` but here I'm importing directly by names so it has no effect

File: sktime/annotation/igts.py (hunk 1 above)
Comment (lmmentel) [reply sktime/annotation/igts.py:18]: It is also present in the class docstrings

File: sktime/annotation/igts.py (hunk 5 above)
Comment (lmmentel) [reply sktime/annotation/igts.py:312]: I think it's more readable with a line break

File: sktime/annotation/igts.py (hunk 8 above)
Comment (lmmentel) [reply sktime/annotation/igts.py:351]: Thanks, but I hope this is non blocking

**Comment by lmmentel** [comment]
`test_constructor` tests if you are exactly doing the `self.ags=arg` in ` __init__` which is testing exactly **one** way of achieving that result in that sense it is testing the implementation syntax and not function and therefore should probably be dropped or allowed to skip. I'm setting the attributes with attrs which in my opinion is superior and cleaner.

**Comment by fkiraly** [comment]
> `test_constructor` tests if you are exactly doing the `self.arg=arg` in ` __init__` 

I don't think that's accurate:

* the test is not looking at the code but the result.
* the tests is testing more, e.g., signature expectations. See the improved docstring in #3555 for an explanation.

> I'm setting the attributes with attrs which in my opinion is superior and cleaner.

Yes, but I believe that breaks the signature expectations that `sklearn`/`sktime` have which are also tested, which is why the test fails.

It may be superior or cleaner but it seems to break `sklearn` compatibility.

**Comment by lmmentel** [comment]
> > `test_constructor` tests if you are exactly doing the `self.arg=arg` in ` __init__`
> 
> I don't think that's accurate:
> 
>     * the test is not looking at the code but the result.
> 
>     * the tests is testing more, e.g., signature expectations. See the improved docstring in [[ENH] improve base class test docstrings and clone test #3555](https://github.com/sktime/sktime/pull/3555) for an explanation.
> 
> 
> > I'm setting the attributes with attrs which in my opinion is superior and cleaner.
> 
> Yes, but I believe that breaks the signature expectations that `sklearn`/`sktime` have which are also tested, which is why the test fails.
> 
> It may be superior or cleaner but it seems to break `sklearn` compatibility.

Nope, it only breaks the test because how the test it structured.

File: sktime/annotation/igts.py [hunk 13]
```diff
@@ -0,0 +1,481 @@
...
+import pandas as pd
+from attrs import asdict, define, field
+
+from sktime.base import BaseEstimator
+
+__all__ = ["InformationGainSegmentation"]
+__author__ = ["lmmentel"]
+
+
+
```
Comment (lmmentel) [inline sktime/annotation/igts.py:35]:
```suggestion
```

**Comment by fkiraly** [comment]
> Nope, it only breaks the test because how the test it structured.

That's contradicting me without explanation/reasoning. I gave you my explanation and reasoning, for a productive discussion we should exchange arguments.

Here's the code, verbatim, that checks init attributes:

```python
        # Ensure that each parameter is set in init
        init_params = _get_args(type(estimator).__init__)
        invalid_attr = set(init_params) - set(vars(estimator)) - {"self"}
```

this looks like it is checking the results to me, against signature conventions.

These are also three lines in a test that is ca 70 lines long, so the implicit claim above that this is all the test does is also incorrect (you can look up that the 67 other lines are contentful).

**Comment by lmmentel** [comment]
You haven't mentioned how exactly my implementation breaks sktime/sklearn compatibility so I'm not sure why would you accuse me of making this discussion unproductive.

**Comment by fkiraly** [comment]
>  so I'm not sure why would you accuse me of making this discussion unproductive.

apologies if I wasn't clear here, I'm not accusing you of anything.

> You haven't mentioned how exactly my implementation breaks sktime/sklearn compatibility

I don't think that's correct, self-citing myself above:

> Yes, but I believe that breaks the signature expectations that `sklearn`/`sktime` have which are also tested, which is why the test fails.

> The tests above test specific interface assumptions required for `sklearn` compatibility, failing them will typically make your estimators not `sklearn` compatible.

> I've added some explanations to the respective tests here:
https://github.com/sktime/sktime/pull/3555
to explain why these tests are important and how to comply with them.
Hope that is useful?

**Comment by lmmentel** [comment]
Writing the words "specific assumptions" does not reveal much about what you have in mind. Also this seems like a nitpick. Without a specific example of what assumption is violated this discussion is pointless.

**Comment by fkiraly** [comment]
well, that's precisely why I made PR #3555 and mentioned it in the discussion above multiple times, trying to explain that it contains explanations.

But for your convenience I'm happy to copy-paste the relevant parts:

```
        """Check that the constructor has sklearn compatible signature and behaviour.

        Based on sklearn check_estimator testing of __init__ logic.
        Uses create_test_instance to create an instance.
        Assumes test_create_test_instance has passed and certified create_test_instance.

        Tests that:
        * constructor has now varargs
        * tests that constructor constructs an instance of the class
        * tests that all parameters are set in init to an attribute of the same name
        * tests that parameter values are always copied to the attribute and not changed
        * tests that default parameters are one of the following:
            None, str, int, float, bool, tuple, function, joblib memory, numpy primitive
            (other type parameters should be None, default handling should be by writing
            the default to attribute of a different name, e.g., my_param_ not my_param)
        """
```

**Comment by fkiraly** [comment]
> Without a specific example of what assumption is violated this discussion is pointless.

What kind of specific example do you want? The new class fails the test. How does this need an example, besids the new class?

**Comment by miraep8** [comment]
Hey all I think I actually ran into a similar issue with the init check when using `define`/`attrs`.

With `attrs` (and I think with what you are doing here @lmmentel), vars seems not to work properly. I think this has something to do with it not playing well with the `__dict__` function that `vars` seems to rely on. However the attributes are correctly added as expected during  `__init__`, (which is captured by using `dir` instead), and I think is what we are really interested in???

`vars` and `dir` are  of course slightly different functions though, as is summarized a bit [here](https://stackoverflow.com/questions/980249/difference-between-dir-and-vars-keys-in-python). 

Basically I was able to get the tests to pass when I made the following change to the init tests:
```
init_params = _get_args(type(estimator).__init__)
invalid_attr = set(init_params) - set(vars(estimator)) - {"self"}
```

change to:

```
init_params = _get_args(type(estimator).__init__)
invalid_attr = set(init_params) - set(dir(estimator)) - {"self"}
```

I ended up not implementing this change in part because @fkiraly was not a fan of the change to the sktime interface tests, but I think since it has come up twice now I think it is worth re-evaluating if making this small change (ie `vars` -> `dir`) is consistent with what the test is looking for.  I personally think it is, and I also think it would fix this problem?  Thoughts?

**Review by miraep8 — COMMENTED** [review]
Just adding a few random comments based on quick impressions.  I may return with others later, overall impressions are that it looks good though and my suggestions are not meant to be blocking.

File: sktime/annotation/igts.py [hunk 14]
```diff
@@ -0,0 +1,486 @@
...
+from sktime.base import BaseEstimator
+
+__all__ = ["InformationGainSegmentation"]
+__author__ = ["lmmentel"]
+
+
+# logger = logging.getLogger(__name__)
+
+
+@dataclass
```
[the remark this answers is not in the document]
Comment (lmmentel) [reply sktime/annotation/igts.py:35]: You are right attrs and dataclasses have similarities on the definition side but for me they serve different purposes. attrs serves to reduce boilerplate for classes and has methods for controlling and tuning that, whereas dataclasses are a data container for moving structured objects around - before dataclasses I would use sth like namedtuple or a dict but a dataclass combines the strengths of both. Hope that helps.

File: sktime/annotation/igts.py [hunk 15]
```diff
@@ -0,0 +1,486 @@
...
+
+def entropy(X: npt.ArrayLike) -> float:
+    """Shannons's entropy for time series.
+
+    As defined by equations (3) and (4) from [1]_.
+
+    Parameters
+    ----------
+    X: array_like
+        Time series data as a 2D numpy array with sequence index along rows
```
[the remark this answers is not in the document]
Comment (lmmentel) [reply sktime/annotation/igts.py:50]: Not sure there was anything explicit in the paper about that, I would have to go in and check but the method won't produce anything sensible for univariate series.

File: sktime/annotation/igts.py [hunk 16]
```diff
@@ -0,0 +1,486 @@
...
+
+def generate_segments(X: npt.ArrayLike, change_points: List[int]) -> npt.ArrayLike:
+    """Generate separate segments from time series based on change points."""
+    for start, end in zip(change_points[:-1], change_points[1:]):
+        yield X[start:end, :]
+
+
+def generate_segments_pandas(X: npt.ArrayLike, change_points: List) -> npt.ArrayLike:
+    """Generate separate segments from time series based on change points."""
+    for interval in pd.IntervalIndex.from_breaks(sorted(change_points), closed="both"):
```
[the remark this answers is not in the document]
Comment (lmmentel) [reply sktime/annotation/igts.py:103]: I thought that `yield` make more sense especially for larger time series. Thanks for the comment, I updated the docstrings.

**Comment by lmmentel** [comment]
> Hey all I think I actually ran into a similar issue with the init check when using `define`/`attrs`.
> 
> With `attrs` (and I think with what you are doing here @lmmentel), vars seems not to work properly. I think this has something to do with it not playing well with the `__dict__` function that `vars` seems to rely on. However the attributes are correctly added as expected during `__init__`, (which is captured by using `dir` instead), and I think is what we are really interested in???
> 
> `vars` and `dir` are of course slightly different functions though, as is summarized a bit [here](https://stackoverflow.com/questions/980249/difference-between-dir-and-vars-keys-in-python).
> 
> Basically I was able to get the tests to pass when I made the following change to the init tests:
> 
> ```
> init_params = _get_args(type(estimator).__init__)
> invalid_attr = set(init_params) - set(vars(estimator)) - {"self"}
> ```
> 
> change to:
> 
> ```
> init_params = _get_args(type(estimator).__init__)
> invalid_attr = set(init_params) - set(dir(estimator)) - {"self"}
> ```
> 
> I ended up not implementing this change in part because @fkiraly was not a fan of the change to the sktime interface tests, but I think since it has come up twice now I think it is worth re-evaluating if making this small change (ie `vars` -> `dir`) is consistent with what the test is looking for. I personally think it is, and I also think it would fix this problem? Thoughts?

Thanks for sharing. My point was a bit broader than fixing the test. I would say that this test serves little purpose other than slowing things down. I have searched and I couldn't find sklearn checking similar aspects so it seems sktime is more dogmatic in this area which by necessity reduces design flexibility but not in way that's desirable.

**Review by aiwalter — APPROVED** [review]

**Pull request merged**